src/librustc_borrowck/borrowck/mir/dataflow/mod.rs

```diff
@@ -0,0 +1,1168 @@
+        let offset = self.words_per_block * block_idx;
+        let range = offset..(offset + self.words_per_block);
+        BlockSets {
+            on_entry: &mut self.on_entry_sets.bits[range.clone()],
+            gen_set: &mut self.gen_sets.bits[range.clone()],
+            kill_set: &mut self.kill_sets.bits[range],
+        }
+    }
+
+    fn lookup_set_for<'a>(&self, sets: &'a Bits, block_idx: usize) -> &'a [usize] {
+        let offset = self.words_per_block * block_idx;
+        &sets.bits[offset..(offset + self.words_per_block)]
+    }
+    pub fn gen_set_for(&self, block_idx: usize) -> &[usize] {
+        self.lookup_set_for(&self.gen_sets, block_idx)
+    }
+    pub fn kill_set_for(&self, block_idx: usize) -> &[usize] {
+        self.lookup_set_for(&self.kill_sets, block_idx)
+    }
+    pub fn on_entry_set_for(&self, block_idx: usize) -> &[usize] {
+        self.lookup_set_for(&self.on_entry_sets, block_idx)
+    }
+}
+
+impl<O: BitDenotation> DataflowState<O> {
+    fn each_bit<F>(&self, ctxt: &O::Ctxt, words: &[usize], mut f: F)
+        where F: FnMut(usize) {
+        //! Helper for iterating over the bits in a bitvector.
+
+        let bits_per_block = self.operator.bits_per_block(ctxt);
+        let usize_bits: usize = mem::size_of::<usize>() * 8;
+
+        for (word_index, &word) in words.iter().enumerate() {
+            if word != 0 {
+                let base_index = word_index * usize_bits;
+                for offset in 0..usize_bits {
+                    let bit = 1 << offset;
+                    if (word & bit) != 0 {
+                        // NB: we round up the total number of bits
+                        // that we store in any given bit set so that
+                        // it is an even multiple of usize::BITS. This
+                        // means that there may be some stray bits at
+                        // the end that do not correspond to any
+                        // actual value; that's why we first check
+                        // that we are in range of bits_per_block.
+                        let bit_index = base_index + offset as usize;
+                        if bit_index >= bits_per_block {
+                            return;
+                        } else {
+                            f(bit_index);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    pub fn interpret_set<'c>(&self, ctxt: &'c O::Ctxt, words: &[usize]) -> Vec<&'c O::Bit> {
+        let mut v = Vec::new();
+        self.each_bit(ctxt, words, |i| {
+            v.push(self.operator.interpret(ctxt, i));
+        });
+        v
+    }
+}
+
+pub trait BitwiseOperator {
+    /// Joins two predecessor bits together, typically either `|` or `&`
+    fn join(&self, pred1: usize, pred2: usize) -> usize;
+}
+
+/// Parameterization for the precise form of data flow that is used.
+pub trait DataflowOperator : BitwiseOperator {
```

Comment:
i agree, it was a holdover from the old dataflow code but it really doesnt make sense to separate the two operations; they are intrisically linked